### ISSUE TITLE
CheckListBox has Bugs

### ISSUE THREAD

**Issue by why5684784** [dotnet/winforms#3892]
* .NET Core Version: 
This is a problem with the. Net framework and does not test the. Net core

* Have you experienced this same bug with .NET Framework?: 
Yes

**Problem description:**
This is a bug from CheckListBox.
When I derive a new CheckListBox from CheckListBox, using my derived MyCheckListBox, it will be very slow to change the itemcheckstate when the item is not selected. It will be very fast to change the itemcheckstate after selecting the item. 

**Expected behavior:**
The rate at which itemcheckstate changes is independent of the itemselect state

**Minimal repro:**
every time

Code:
`
```
using System;
using System.Windows.Forms;
namespace Test.Frm.Control
{
    internal class MyCheckListBox : CheckedListBox
    {
        public delegate void ItemBoxCheckedEventHanlder(object sender, Models.CheckListBoxItemCheckBoxEventArgs mouseEventArgs);
        public event ItemBoxCheckedEventHanlder ItemBoxChecked;
        private bool isUserCheck = false;
        public MyCheckListBox()
        {
            this.DoubleBuffered = false;
        }
        protected override void OnItemCheck(ItemCheckEventArgs ice)
        {
            if (this.isUserCheck)
            {
                this.isUserCheck = false;
                ice.NewValue = ice.CurrentValue;
            }
            base.OnItemCheck(ice);
        }
        protected override void WndProc(ref Message m)
        {
            if (m.Msg == 0x0201)///LeftMouseButtonDown
            {
                uint lPararm = (uint)m.LParam;
                int xPoint = Convert.ToInt32(lPararm & 0xFFFF);///Convert mouse X coordinates
                int yPoint = Convert.ToInt32(lPararm >> 16);///Convert mouse Y coordinates
                int index = this.IndexFromPoint(xPoint, yPoint);
                if (xPoint < 15)
                {
                    Models.CheckListBoxItemCheckBoxEventArgs model = new Models.CheckListBoxItemCheckBoxEventArgs(MouseButtons.Left, xPoint, yPoint, index);
                    base.SetItemCheckState(index, base.GetItemCheckState(index) == CheckState.Unchecked ? CheckState.Checked : CheckState.Unchecked);
                    this.ItemBoxChecked?.Invoke(this, model);
                }
                else
                {
                    this.isUserCheck = true;
                    base.SelectedIndex = index;
                }
            }
            else
            {
                base.WndProc(ref m);
            }
        }
    }
}
```
`

**Comment by RussKie** [comment]
> * This is a problem with the. Net framework and does not test the. Net core

Thank you for the submission, but this repo is for .NET Core/.NET versions only. If you're welcome to update the issue if you think it is a bug in one of those. It would be great to have an animation that shows the issue.

All bugs for .NET Frameworks should be raised via VS Feedback. But I will be upfront with you - .NET Framework is done, and only critical issues will meet the servicing bar. And I can't see this bug meeting the bar.

**Comment by why5684784** [comment]
This problem extends from. Net framework to. Net core, and still exists in various versions.

**Comment by elachlan** [comment]
@why5684784 I can't reproduce this. Can you please provide a more complete simple reproduction?

**Comment by elachlan** [comment]
@Olina-Zhang I was unable to reproduce this bug, could your team give it a go. If we can't reproduce it, we can close this issue.

**Comment by Olina-Zhang** [comment]
@elachlan This issue can reproduce in both .Net core and .Net framework for derived MyCheckListBox using following test sample. For derived MyCheckListBox, when focus on item, then move mouse to its checkBox, we need to click twice to check/uncheck. But Winforms standalone CheckedListBox, just do once.
[CheckedListBoxSample.zip](https://github.com/dotnet/winforms/files/10441527/CheckedListBoxSample.zip)
![CheckBoxIssue](https://user-images.githubusercontent.com/26474449/213070563-07d7a59b-f63a-4f87-8179-30bb3f598639.gif)

**Comment by JeremyKuhne** [comment]
@elchan if you're interested in looking into this it would be great, otherwise we're unlikely to get to it.

**Comment by elachlan** [comment]
@JeremyKuhne can you please remove the auto-close on this please?

**Comment by JeremyKuhne** [comment]
@elachlan I presume you're going to look at it then?

**Comment by elachlan** [comment]
That's the plan. Maybe add help wanted so we get prompted later down the line if I don't get to it.

**Comment by joshuanrobinson** [comment]
I don't think this is a bug with `CheckedListBox`. The issue described by @Olina-Zhang is occurring due to the way `MyCheckedListBox` is coded.

I'm making a bit of a guess here, but it looks like `MyCheckedListBox` is designed to only modify an item's checked state if the user interacts with the checkbox itself. If the left mouse button down occurs while the cursor is over any other part of the item then the `else` block runs and `this.isUserCheck` gets set to `true`. However, the check state of the items are not modified so the overwritten `OnItemCheck` never gets called.

The next time the user clicks on an item's checkbox, the `if (xPoint < 15)` block runs, which modifies the check state of the item that the mouse is currently over. This causes the overwritten `OnItemCheck` to run, however the `isUserCheck` field is still set to `true`. This causes the `if (this.isUserCheck)` block to execute, which sets `isUserCheck` to `false` and prevents the item's check state from actually changing. If the user clicks on the checkbox again, the item's check state is changed because `isUserCheck` is `false` when execution enters `OnItemCheck`.
